test/unit/range/view/view_async_buffer_test.cpp

```diff
@@ -0,0 +1,106 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2019, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2019, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+#include <gtest/gtest.h>
+
+#include <iostream>
+#include <string>
+#include <vector>
+
+#include <seqan3/alphabet/nucleotide/all.hpp>
+#include <seqan3/range/concept.hpp>
+#include <seqan3/range/view/async_buffer.hpp>
+#include <seqan3/range/view/single_pass_input.hpp>
+#include <seqan3/std/ranges>
+
+#include "../iterator_test_template.hpp"
+
+using namespace seqan3;
+
+using iterator_type = std::ranges::iterator_t<
+    decltype(std::declval<std::vector<seqan3::dna4>&>() | view::async_buffer(3))>;
+
+template <>
+struct iterator_fixture<iterator_type> : public ::testing::Test
+{
+    using iterator_tag = std::input_iterator_tag;
+    static constexpr bool const_iterable = false;
+
+    std::vector<seqan3::dna4> vec{"ACGTACGTACGTATCGAGAGCTTTAGC"_dna4};
+    std::vector<seqan3::dna4> expected_range{"ACGTACGTACGTATCGAGAGCTTTAGC"_dna4};
+    decltype(view::async_buffer(vec, 3)) test_range = view::async_buffer(vec, 3);
+};
+
+using test_type = ::testing::Types<iterator_type>;
+INSTANTIATE_TYPED_TEST_CASE_P(iterator_fixture, iterator_fixture, test_type);
+
+
+TEST(async_buffer, in_out)
+{
+    std::vector<seqan3::dna4> vec{"ACGTACGTACGTATCGAGAGCTTTAGC"_dna4};
+
+    auto v = vec | view::async_buffer(3);
+
+    EXPECT_TRUE(std::ranges::equal(vec, v));
+}
+
+TEST(async_buffer, buffer_size_zero)
+{
+    std::vector<seqan3::dna4> vec{"ACGTACGTACGTATCGAGAGCTTTAGC"_dna4};
+
+    EXPECT_THROW(vec | view::async_buffer(0), std::invalid_argument);
+}
+
+TEST(async_buffer, buffer_size_huge)
+{
+    std::vector<seqan3::dna4> vec{"ACGTACGTACGTATCGAGAGCTTTAGC"_dna4};
+
+    auto v = vec | view::async_buffer(100000);
+
+    EXPECT_TRUE(std::ranges::equal(vec, v));
+}
+
+TEST(async_buffer, destruct_with_full_buffer)
+{
+    std::vector<seqan3::dna4> vec{"ACGTACGTACGTATCGAGAGCTTTAGC"_dna4};
+
+    auto v0 = vec | view::single_pass_input;
+
+    {
+        auto v1 = v0 | view::async_buffer(5);
+
+        // consume five elements (construction already consumes one)
+        auto b = std::ranges::begin(v1);
+        ++b; ++b; ++b; ++b;
+
+        // give time to rebuffer next five elements
+        std::this_thread::sleep_for(std::chrono::milliseconds(100));
```

Comment:
this is a bad test, since you have no guarantee that the producer thread will rebuffer in this time. It could be yielded and come alive first after the 100 milliseconds or within. I am afraid this can't be solved easily.

Reply:
I need to test exactly that behaviour. I know behaviour is not guaranteed, but I think it's fairly reasonable to expect this to work in 99.9999999% of the cases.

I don't see another way of testing this, short of adding undocumented public accessors to the view that expose the internal "fill-state". That would be **very ugly**.

Reply:
Well, I'd like you to remember the weird scenarios we had with parallel tests we assumed they were correct and still caused a lot of trouble. Now you arguing, a test that is wrong by design is ok since it presumably happens not very often? I don't understand the reasoning. As said before you need to handle it differently. In this case I guess what you need is a barrier and on destruction of the iterator you arrive on the barrier within the buffer. If all iterators arrived (destructed) you can safely destruct the buffer after closing it and waiting for the producer thread. That is the only way to prevent a data race.